File: app/src/main/java/piuk/blockchain/android/ui/backup/BackupWalletWordListFragment.java
package piuk.blockchain.android.ui.backup;

import android.app.Fragment;
import android.app.FragmentTransaction;
import android.content.Context;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.view.inputmethod.InputMethodManager;

import java.util.List;
import piuk.blockchain.android.R;
import piuk.blockchain.android.databinding.FragmentBackupWordListBinding;
import piuk.blockchain.android.util.BackupWalletUtil;
import piuk.blockchain.android.util.annotations.Thunk;

public class BackupWalletWordListFragment extends Fragment {

    @Thunk FragmentBackupWordListBinding binding;
    @Thunk Animation animEnterFromRight;
    @Thunk Animation animEnterFromLeft;
    private Animation animExitToLeft;
    private Animation animExitToRight;

    int currentWordIndex = 0;
    @Thunk
    List<String> mnemonic;
    @Thunk String word;
    @Thunk String of;
    private String secondPassword;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_backup_word_list, container, false);

        Bundle bundle = getArguments();
        if (bundle != null) {
            secondPassword = bundle.getString("second_password");
        }

        word = getResources().getString(R.string.backup_word);
        of = getResources().getString(R.string.backup_of);

        animExitToLeft = AnimationUtils.loadAnimation(getActivity(), R.anim.exit_to_left);
        animEnterFromRight = AnimationUtils.loadAnimation(getActivity(), R.anim.enter_from_right);

        animExitToRight = AnimationUtils.loadAnimation(getActivity(), R.anim.exit_to_right);
        animEnterFromLeft = AnimationUtils.loadAnimation(getActivity(), R.anim.enter_from_left);

        mnemonic = new BackupWalletUtil().getMnemonic(secondPassword);
        if (currentWordIndex == mnemonic.size()) {
            currentWordIndex = 0;
        }
        binding.tvCurrentWord.setText(word + " " + (currentWordIndex + 1) + " " + of + " 12");
        binding.tvPressReveal.setText(mnemonic.get(currentWordIndex));

        binding.nextWordAction.setOnClickListener(v -> {

            if (currentWordIndex >= 0) {
                binding.previousWordAction.setVisibility(View.VISIBLE);
            } else {
                binding.previousWordAction.setVisibility(View.INVISIBLE);
            }

            if (currentWordIndex < mnemonic.size() - 1) {

                animExitToLeft.setAnimationListener(new Animation.AnimationListener() {
                    @Override
                    public void onAnimationStart(Animation animation) {
                        binding.tvPressReveal.setText("");
                        binding.tvCurrentWord.setText(word + " " + (currentWordIndex + 1) + " " + of + " 12");
                    }

                    @Override
                    public void onAnimationRepeat(Animation animation) {
                        // No-op
                    }

                    @Override
                    public void onAnimationEnd(Animation animation) {
                        binding.cardLayout.startAnimation(animEnterFromRight);
                        binding.tvPressReveal.setText(mnemonic.get(currentWordIndex));
                    }
                });

                binding.cardLayout.startAnimation(animExitToLeft);

            }

            currentWordIndex++;

            if (currentWordIndex == mnemonic.size()) {

                currentWordIndex = 0;

                Fragment fragment = new BackupWalletVerifyFragment();
                if (secondPassword != null) {
                    Bundle args = new Bundle();
                    args.putString("second_password", secondPassword);
                    fragment.setArguments(args);
                }

                getFragmentManager().beginTransaction()
                        .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                        .replace(R.id.content_frame, fragment)
                        .addToBackStack(null)
                        .commit();
            } else {

                if (currentWordIndex == mnemonic.size() - 1) {
                    binding.nextWordAction.setText(getResources().getString(R.string.backup_done));
                } else {
                    binding.nextWordAction.setText(getResources().getString(R.string.backup_next_word));
                }
            }
        });

        binding.previousWordAction.setOnClickListener(v1 -> {

            binding.nextWordAction.setText(getResources().getString(R.string.backup_next_word));

            if (currentWordIndex == 1) {
                binding.previousWordAction.setVisibility(View.INVISIBLE);
            }

            animExitToRight.setAnimationListener(new Animation.AnimationListener() {
                @Override
                public void onAnimationStart(Animation animation) {
                    binding.tvPressReveal.setText("");
                    binding.tvCurrentWord.setText(word + " " + (currentWordIndex + 1) + " " + of + " 12");
                }

                @Override
                public void onAnimationRepeat(Animation animation) {
                    // No-op
                }

                @Override
                public void onAnimationEnd(Animation animation) {
                    binding.cardLayout.startAnimation(animEnterFromLeft);
                    binding.tvPressReveal.setText(mnemonic.get(currentWordIndex));
                }
            });

            binding.cardLayout.startAnimation(animExitToRight);

            currentWordIndex--;
        });

        return binding.getRoot();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        View view = getActivity().getCurrentFocus();
        if (view != null) {
            InputMethodManager imm = (InputMethodManager) getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
            imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
        }
    }
}

File: app/src/main/java/piuk/blockchain/android/ui/backup/wordlist/BackupWalletWordListFragment.kt
package piuk.blockchain.android.ui.backup.wordlist

import android.app.FragmentTransaction
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.view.inputmethod.InputMethodManager
import kotlinx.android.synthetic.main.fragment_backup_word_list.*
import piuk.blockchain.android.R
import piuk.blockchain.android.injection.Injector
import piuk.blockchain.android.ui.backup.verify.BackupWalletVerifyFragment
import piuk.blockchain.android.ui.base.BaseFragment
import piuk.blockchain.android.util.extensions.inflate
import piuk.blockchain.android.util.extensions.invisible
import piuk.blockchain.android.util.extensions.visible
import piuk.blockchain.android.util.helperfunctions.unsafeLazy
import javax.inject.Inject

class BackupWalletWordListFragment : BaseFragment<BackupWalletWordListView, BackupWalletWordListPresenter>(),
        BackupWalletWordListView {

    @Inject lateinit var backupWalletWordListPresenter: BackupWalletWordListPresenter

    private val animEnterFromRight: Animation by unsafeLazy { AnimationUtils.loadAnimation(activity, R.anim.enter_from_right) }
    private val animEnterFromLeft: Animation by unsafeLazy { AnimationUtils.loadAnimation(activity, R.anim.enter_from_left) }
    private val animExitToLeft: Animation by unsafeLazy { AnimationUtils.loadAnimation(activity, R.anim.exit_to_left) }
    private val animExitToRight: Animation by unsafeLazy { AnimationUtils.loadAnimation(activity, R.anim.exit_to_right) }
    private val word: String by unsafeLazy { getString(R.string.backup_word) }
    private val of: String by unsafeLazy { getString(R.string.backup_of) }

    var currentWordIndex = 0

    init {
        Injector.getInstance().presenterComponent.inject(this)
    }

    override fun onCreateView(
            inflater: LayoutInflater,
            container: ViewGroup?,
            savedInstanceState: Bundle?
    ): View? = container?.inflate(R.layout.fragment_backup_word_list)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        onViewReady()

        textview_word_counter.text = getFormattedPositionString()
        textview_current_word.text = presenter.getWordForIndex(currentWordIndex)

        button_next.setOnClickListener { onNextClicked() }
        button_previous.setOnClickListener { onPreviousClicked() }
    }

    override fun getPageBundle(): Bundle? = arguments

    override fun createPresenter() = backupWalletWordListPresenter

    override fun getMvpView() = this

    override fun finish() {
        activity?.finish()
    }

    private fun onNextClicked() {
        if (currentWordIndex >= 0) button_previous.visible() else button_previous.invisible()

        if (currentWordIndex < presenter.getMnemonicSize() - 1) {
            animExitToLeft.setAnimationListener(object : Animation.AnimationListener {
                override fun onAnimationStart(animation: Animation) {
                    textview_current_word.text = ""
                    textview_word_counter.text = getFormattedPositionString()
                }

                override fun onAnimationRepeat(animation: Animation) {
                    // No-op
                }

                override fun onAnimationEnd(animation: Animation) {
                    card_layout?.startAnimation(animEnterFromRight)
                    textview_current_word?.text = presenter.getWordForIndex(currentWordIndex)
                }
            })

            card_layout.startAnimation(animExitToLeft)
        }

        currentWordIndex++

        if (currentWordIndex == presenter.getMnemonicSize()) {
            currentWordIndex = 0
            launchVerifyFragment()
        } else {
            if (currentWordIndex == presenter.getMnemonicSize() - 1) {
                button_next.text = getString(R.string.backup_done)
            }
        }
    }

    private fun onPreviousClicked() {
        button_next.text = getString(R.string.backup_next_word)
        if (currentWordIndex == 1) {
            button_previous.invisible()
        }

        animExitToRight.setAnimationListener(object : Animation.AnimationListener {
            override fun onAnimationStart(animation: Animation) {
                textview_current_word.text = ""
                textview_word_counter.text = getFormattedPositionString()
            }

            override fun onAnimationRepeat(animation: Animation) {
                // No-op
            }

            override fun onAnimationEnd(animation: Animation) {
                card_layout?.startAnimation(animEnterFromLeft)
                textview_current_word?.text = presenter.getWordForIndex(currentWordIndex)
            }
        })

        card_layout.startAnimation(animExitToRight)
        currentWordIndex--
    }

    private fun launchVerifyFragment() {
        val fragment = BackupWalletVerifyFragment()
        if (presenter.secondPassword != null) {
            fragment.arguments = Bundle().apply {
                putString(ARGUMENT_SECOND_PASSWORD, presenter.secondPassword)
            }
        }

        fragmentManager?.run {
            beginTransaction()
                    .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                    .replace(R.id.content_frame, fragment)
                    .addToBackStack(null)
                    .commit()
        }
    }

    private fun getFormattedPositionString(): CharSequence? = "$word ${currentWordIndex + 1} $of 12"

    override fun onDestroyView() {
        super.onDestroyView()
        activity?.run {
            val view = currentFocus
            if (view != null) {
                val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                imm.hideSoftInputFromWindow(view.windowToken, 0)
            }
        }
    }

    companion object {
        const val ARGUMENT_SECOND_PASSWORD = "second_password"
    }
}